File: cliente.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include "config.h"

#define CONFIG_FILE "config.conf"
#define SIGNOS_FILE "signos.txt"

int main(int argc, char *argv[]){

    Config cfg;
    int sock_fd;
    struct sockaddr_in serv_addr;
    char *buffer;
    char signo[50], fecha[50];

    if (cargar_configuracion(CONFIG_FILE, SIGNOS_FILE, NULL, &cfg) < 0) {
        fprintf(stderr, "Error al cargar configuracion\n");
        exit(1);
    }

    buffer = malloc(cfg.tamano_buffer);
    if (!buffer) {
        perror("Error al asignar memoria");
        exit(1);
    }

    if(argc != 3) {
        fprintf(stderr, "Entrada esperada: %s <signo> <dd/mm/yy>\n", argv[0] );
        fprintf(stderr, "Ejemplo: %s leo <12/08/1999>\n", argv[0] );
        exit(1);
    }

    strncpy(signo, argv[1], sizeof(signo) - 1);
    strncpy(fecha, argv[2], sizeof(fecha) - 1);

    sock_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (sock_fd < 0) {
        perror("Error al crear socket");
        exit(1);
    }

    memset(&serv_addr, 0, sizeof(serv_addr));
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_addr.s_addr  = inet_addr(cfg.ip_central);
    serv_addr.sin_port = htons(cfg.puerto_central);

    if(connect(sock_fd, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0){
        perror("Error al intentar conectarse con el servidor central");
    }

     printf("Conectado al Servidor Central (%s:%d)\n", cfg.ip_central, cfg.puerto_central);
    
    char clave[100];
    snprintf(clave, sizeof(clave), "%s|%s", signo, fecha);

    strncpy(buffer, clave, sizeof(clave) - 1);

    if (send(sock_fd, buffer, strlen(buffer), 0) < 0) {
        perror("Error al enviar datos");
        exit(1);
    }

     printf("Consulta enviada: Signo=%s, Fecha=%s\n", signo, fecha);

    memset(buffer, 0, cfg.tamano_buffer);
    if (recv(sock_fd, buffer, cfg.tamano_buffer - 1, 0) < 0) {
        perror("Error al recibir respuesta");
        exit(1);
    }

    printf("\n===== RESULTADOS =====\n");
    printf("%s\n", buffer);
    printf("======================\n");
    
    free(buffer);
    close(sock_fd);
    return 0;

}

File: LAB2/cliente/Cliente.java
import java.io.IOException;
import java.rmi.registry.LocateRegistry;
import java.rmi.registry.Registry;

public class Cliente {
    public static void main(String[] args) {
        if (args.length != 2) {
            System.err.println("Entrada esperada: java Cliente <signo> <dd/mm/yyyy>");
            System.err.println("Ejemplo: java Cliente leo 12/08/1999");
            return;
        }

        try {
            Config config = Config.load("config.conf");
            Registry registry = LocateRegistry.getRegistry(config.ipCentral, config.puertoCentral);
            ConsultaCentralRemote remote = (ConsultaCentralRemote) registry.lookup(ConsultaCentralRemote.SERVICE_NAME);

            ConsultaRequest request = new ConsultaRequest(args[0], args[1]);
            ConsultaResponse response = remote.consultar(request);

            System.out.println("Conectado al Servidor Central (" + config.ipCentral + ":" + config.puertoCentral + ")");
            System.out.println("Consulta enviada: Signo=" + request.getSigno() + ", Fecha=" + request.getFecha());
            System.out.println();
            System.out.println("===== RESULTADOS =====");
            System.out.println(response.format());
            if (response.isDesdeCache()) {
                System.out.println("(respuesta obtenida desde cache)");
            }
            System.out.println("======================");
        } catch (IOException exception) {
            System.err.println("No se pudo cargar la configuracion del cliente: " + exception.getMessage());
        } catch (Exception exception) {
            System.err.println("Error al consultar el Servidor Central: " + exception.getMessage());
        }
    }
}
